[DRMAtomic] Setting DRM property by name instead of by hardcoded enum value

slattann opened the pull request:
This patch implements setting DRM property like Scaling_Filter by
validating actual supported values than by hard coding through local
enums.
 1) Added new API 'AddPropertyValueByName'
 2) removed the enum defitiion that hard codes the property values

Signed-off-by: Sameer Lattannavar <sameer.lattannavar@intel.com>

## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

ppaalanen commented on xbmc/windowing/gbm/drm/DRMAtomic.cpp:
```diff
@@ -267,6 +249,32 @@ bool CDRMAtomic::AddProperty(CDRMObject* object, const char* name, uint64_t valu
+{
+  for (uint32_t i = 0; i < object->props->count_props; i++)
+  {
+    if (!strcmp(object->props_info[i]->name, name))
```
This might be easier to read if you used `continue` when the name does not match. The same for the other `strcmp()` below.

ppaalanen commented on xbmc/windowing/gbm/drm/DRMAtomic.cpp:
```diff
@@ -267,6 +249,32 @@ bool CDRMAtomic::AddProperty(CDRMObject* object, const char* name, uint64_t valu
+      }
+    }
+  }
+  return true;
```
You probably do not want to return `true` if the property or the enum value you are trying to set does not exist.

ppaalanen commented on xbmc/windowing/gbm/drm/DRMAtomic.cpp:
```diff
@@ -60,22 +46,18 @@ uint32_t GetScalingFactor(uint32_t srcWidth,
+    AddProperty(object, "CRTC_X", (diff_w / 2));
+    AddProperty(object, "CRTC_Y", (diff_h / 2));
+    AddProperty(object, "CRTC_W", (mar_scale_factor * m_width));
+    AddProperty(object, "CRTC_H", (mar_scale_factor * m_height));
```
Btw. what will happen if DRM rejects these CRTC coordinates? That is, KMS atomic commit fails.

Not all planes might support values different from (0, 0, hdisplay, vdisplay).

slattann replied:
The idea here is to match the way GL is doing to make frame centralize after upscaling. I am not sure if there are some planes do not allow to request drawing from position other than (0, 0).

ppaalanen replied:
Traditionally all primary planes refuse if you don't fill the whole plane exactly with the FB.

Any driver could refuse it for any reason, so there needs to be a fallback, but that applies to atomic KMS in general.

slattann replied:
I got it. With that change user experience may not be good I thought. Upscaled frame will be drawn on left top corner and right bottom corner will be painted black.
Attached the difference in the user experience for your reference. 
![image](https://user-images.githubusercontent.com/25810508/96134674-f50f8980-0f18-11eb-9a76-34a8c192995a.png)
I felt right one is better in viewing the game content so I have added that code. 
We can do one check if 
` AddProperty(object, "CRTC_X", (diff_w / 2));
  AddProperty(object, "CRTC_Y", (diff_h / 2));` 
fail then we can set to (0, 0).  What do you suggest ?

slattann replied:
ok. I will keep  CRTC (X, Y) to (0, 0) as it was before.

ppaalanen replied:
No, that is not what I meant.

Setting the CRTC X, Y, W, H to center the integer-scaled image is fine!

However, you **must** prepare for the fact that the KMS driver may not be able to do that, so you need a *fallback* path.

ppaalanen reviewed (changes requested):
A quick look on this patch in isolation, that is, not knowing anything about the surrounding code, gives me the impression that this is correct usage of the KMS UAPI.

ppaalanen commented:
What hardware and driver have you tested this on?

slattann commented:
> What hardware and driver have you tested this on?

Tested it on TigerLake + i915 driver.

lrusak commented on xbmc/windowing/gbm/drm/DRMAtomic.cpp:
```diff
@@ -267,6 +245,34 @@ bool CDRMAtomic::AddProperty(CDRMObject* object, const char* name, uint64_t valu
   return true;
 }
 
+bool CDRMAtomic::AddPropertyValueByName(struct drm_object* object,
```
`drm_object` struct doesn't exist anymore. You'll have to use the `CDRMObject` class.

slattann replied:
Ok got it. I will fix it .

lrusak commented:
I think something like this would be helpful here (untested)


```c++

bool CDRMObject::GetPropertyValue(std::string name, std::string type, uint64_t& value) const
{
  auto property = std::find_if(m_propsInfo.begin(), m_propsInfo.end(), [&name](auto& prop) {
    return StringUtils::EqualsNoCase(prop->name, name);
  });

  if (property == m_propsInfo.end())
    return false;

  if (drm_property_type_is(property->get(), DRM_MODE_PROP_ENUM) != 0)
  {
    for (int j = 0; j < property->get()->count_enums; j++)
    {
      if (std::strcmp(property->get()->enums[j].name, type.c_str()) == 0)
      {
        value = property->get()->enums[j].value;
        return true;
      }
    }
  }

  return false;
}
```

use with
```c++
uint64_t value;
if (m_gui_plane->GetPropertyValue("SCALING_FILTER", "Nearest Neighbor", value);
  AddProperty(m_gui_plane, "SCALING_FILTER", value);
```

or something similar

slattann committed:
[DRMAtomic] Setting DRM property by name instead of by local enums

This patch implements setting DRM property like Scaling_Filter by
validating actual supported values than by hard coding through local
enums.
 1) Added new API 'GetPropertyValue'
 2) removed the enum defitiion that hard codes the property values

Addressed review comment from Pekka.
Removed the centralizing logic to avoid if any planes fail to set
non-zero integers to CRTC (X,Y).
Addressing 'CDRMObject' instead of 'struct drm_object'.
Fixed an error in the code. Thanks to Pekka for pointing it out.

Signed-off-by: Sameer Lattannavar <sameer.lattannavar@intel.com>
Signed-off-by: Lukas Rusak <lorusak@gmail.com>

slattann commented:
> I think something like this would be helpful here (untested)
> 
> ```c++
> bool CDRMObject::GetPropertyValue(std::string name, std::string type, uint64_t& value) const
> {
>   auto property = std::find_if(m_propsInfo.begin(), m_propsInfo.end(), [&name](auto& prop) {
>     return StringUtils::EqualsNoCase(prop->name, name);
>   });
> 
>   if (property == m_propsInfo.end())
>     return false;
> 
>   auto prop = property->get();
> 
>   if (drm_property_type_is(prop, DRM_MODE_PROP_ENUM) != 0)
>   {
>     for (int j = 0; j < prop->count_enums; j++)
>     {
>       if (std::strcmp(prop->enums[j].name, type.c_str()) == 0)
>       {
>         value = prop->enums[j].value;
>         return true;
>       }
>     }
>   }
> 
>   return false;
> }
> ```
> 
> use with
> 
> ```c++
> uint64_t value;
> if (m_gui_plane->GetPropertyValue("SCALING_FILTER", "Nearest Neighbor", value))
>   AddProperty(m_gui_plane, "SCALING_FILTER", value);
> ```
> 
> or something similar
> 
> EDIT:
> are properties allowed to be 0? if not I would rather return the property value instead of bool and check `value != 0`

@lrusak, I have tested this for getting the right scaling_filter property value and it works. I have updated this PR. Kindly help to get this merged. Kernel Patches are pending to merge because of this blocker. Thanks

lrusak reviewed (approved):
LGTM

lrusak commented:
jenkins build this please

slattann commented:
Once again thank you so much to @ppaalanen  @lrusak  & @garbear  to help address everything to do right.

ppaalanen commented:
> ```c++
> StringUtils::EqualsNoCase
> ```

Does that mean case-insensitive match? DRM property names are case-sensitive. And yes, there is no common pattern for using upper-case letters in the names, so you do want to match the kernel spelling of a name literally.

0 is always an invalid id with KMS, so 0 cannot be a valid property id. Property values in general are allowed to be 0. It looks like 0 is a valid enum value:

> DRM_MODE_PROP_ENUM
>    Enumerated properties take a numerical value that ranges from 0 to the number of enumerated values defined by the property minus one, and associate a free-formed string name to each value. Applications can retrieve the list of defined value-name pairs and use the numerical value to get and set property instance values.

from https://www.kernel.org/doc/html/latest/gpu/drm-kms.html#property-types-and-blob-property-support

Otherwise that example code from @lrusak looks good for what I can understand of C++, though maybe `GetPropertyValue` might be confused with getting a property's current value from KMS rather than getting the uint64 corresponding to an enum value name?

ppaalanen commented on xbmc/windowing/gbm/drm/DRMAtomic.cpp:
```diff
@@ -60,17 +46,13 @@ uint32_t GetScalingFactor(uint32_t srcWidth,
 {
-  uint32_t filter_type = GetScalingFilterType(type);
-  if (object->SupportsPropertyAndValue(name, filter_type))
+  uint64_t filter_type = GetScalingFilterType(type);
```
Does this even build? How can you say you tested this if it doesn't build?

slattann replied:
My bad, some garbage code left before pushing into PR, I will fix it. Thanks for pointing it out.

ppaalanen commented on xbmc/windowing/gbm/drm/DRMAtomic.cpp:
```diff
@@ -60,17 +46,13 @@ uint32_t GetScalingFactor(uint32_t srcWidth,
-      AddProperty(object, "CRTC_X", (diff_w / 2));
-      AddProperty(object, "CRTC_Y", (diff_h / 2));
       AddProperty(object, "CRTC_W", (mar_scale_factor * m_width));
       AddProperty(object, "CRTC_H", (mar_scale_factor * m_height));
```
As I said, setting the CRTC X, Y, W, H to center the integer-scaled image is fine. You just need to prepare for the fact that the KMS driver may not support that, particularly if the KMS plane happens to be the primary plane.

If that does not work, the it is very likely that you cannot have `"CRTC_W", (mar_scale_factor * m_width)` or `"CRTC_H", (mar_scale_factor * m_height)` either. How do you handle it when the KMS driver refuses?

It looks like `CDRMAtomic::DrmAtomicCommit()` has literally no fallback plan. It attempts a test-only commit, yes, but if that fails, you are supposed to try something else until it succeeds. Right now it looks like the user would be left with a black or frozen screen.

So you would try with the centered integer-scaled image first. If that fails, then must try again without integer scaling.

ppaalanen replied:
Please, explain how this review comment has been addressed, or alternatively why it does not need addressing?

emersion reviewed (approved):
This looks like correct kernel uAPI usage to me. Thanks for fixing this!

lrusak commented:
jenkins build this please

slattann commented:
Since Kernel patches have been merged into [drm-tip](https://github.com/freedesktop/drm-tip/commit/5c759eda9b04cd1047b4cda8ac1eaadf8a9e4fce), can we please merge this PR?

howie-f commented:
jenkins build this please

howie-f commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

ppaalanen commented:
@slattann, since your comment sounds like you believe the kernel patch landed, let me ask you this.

Quoting https://www.kernel.org/doc/html/latest/gpu/drm-uapi.html#open-source-userspace-requirements :

> The kernel patch can only be merged after all the above requirements are met, but it must be merged to either drm-next or drm-misc-next before the userspace patches land.

Is drm-tip one of the drm-next or drm-misc-next, or is drm-tip just a throw-away integration-testing branch from which commits can disappear at any time?

emersion commented:
>Is drm-tip one of the drm-next or drm-misc-next, or is drm-tip just a throw-away integration-testing branch from which commits can disappear at any time?

drm-tip is a throw-away integration-testing branch. Each time you push to drm-next or drm-misc-next (or other repos), the drm-tip integration branch is re-generated. More info: https://drm.pages.freedesktop.org/maintainer-tools/drm-tip.html

DaveTBlake commented:
jenkins build this please
@lrusak @garbear you both have approved, one of you like to hit merge (or not, and why) I know nothing about DRM etc.

lrusak commented:
> jenkins build this please
> @lrusak @garbear you both have approved, one of you like to hit merge (or not, and why) I know nothing about DRM etc.

just waiting for a successful jenkins build :)

DaveTBlake commented:
jenkins build this please

garbear commented:
@ppaalanen @emersion @slattann Is now a good time to merge? We're wrapping up PRs into our last alpha. We gladly accept less-that-perfect solutions this early, because we'll have a long time to sort out any changes or fixes that need to be made.

DaveTBlake commented:
@garbear it sounded like @lrusak was ready to mege this, so it would have been merged expects for jenkins being difficult. Once green I'll merge this.

DaveTBlake commented:
Right build separately jenkins CI is happy. See https://jenkins.kodi.tv/view/Helpers/job/BuildMulti-PR-Manually/400/ so hitting that merge button.
